src/Caching/Hybrid/src/Internal/DefaultHybridCache.Stampede.cs

```diff
@@ -0,0 +1,51 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Extensions.Caching.Hybrid.Internal;
+
+partial class DefaultHybridCache
+{
+    private readonly ConcurrentDictionary<StampedeKey, StampedeState> _currentOperations = new();
+
+    internal int DebugGetCallerCount(string key, HybridCacheEntryFlags? flags = null)
+    {
+        var stampedeKey = new StampedeKey(key, flags ?? _defaultFlags);
+        return _currentOperations.TryGetValue(stampedeKey, out var state) ? state.DebugCallerCount : 0;
+    }
+
+    // returns true for a new session (in which case: we need to start the work), false for a pre-existing session
+    public bool GetOrCreateStampede<TState, T>(string key, HybridCacheEntryFlags flags, out StampedeState<TState, T> stampedeState, bool canBeCanceled)
+    {
+        var stampedeKey = new StampedeKey(key, flags);
+        if (_currentOperations.TryGetValue(stampedeKey, out var found))
+        {
+            var tmp = found as StampedeState<TState, T>;
+            if (tmp is null)
+            {
+                ThrowWrongType(key);
+            }
+
+            if (tmp.TryAddCaller())
+            {
+                // we joined an existing session
+                stampedeState = tmp;
+                return false;
+            }
+        }
+
+        // create a new session
+        stampedeState = new StampedeState<TState, T>(this, stampedeKey, canBeCanceled);
```

Comment:
Adding a stampede state isn't thread safe with parallel callers for the same key/flag pair.

Reply:
I was thinking about that overnight; I don't really want to have to `lock` for the final update step (adding a double-check inside the `lock`), but I guess it wouldn't be the end of the world if we can at least do most of the reads without it